### DISCUSSION TITLE
[V3] Difference in update hook when binding to array compared to V2

### DISCUSSION THREAD

**Discussion by AntonioDiPassio-AppSys** [livewire/livewire#6665] [7. Bugs (archived)]
### Livewire version

v3.0.1

### Which PHP version are you using?

PHP 8.1

### Steps To Reproduce

Setup both a livewire 2 and livewire 3 installation:

Livewire 2
1. laravel new lw2
2. cd lw2
3. composer require livewire/livewire:"^2.0"
4. php artisan livewire:make test-checkboxes
5. edit /resources/views/welcome.blade.php
```
<html>
    <head>
        @livewireScripts
    </head>
    <body>
        <livewire:test-checkboxes/>
    </body>
</html>

```
6. edit /app/Http/Livewire/TestCheckboxes.php
```
<?php

namespace App\Http\Livewire;

use Livewire\Component;

class TestCheckboxes extends Component
{
    public $checkboxes = [
        'one',
        'two',
        'three'
    ];

    public $selectedCheckboxes = [
        'one',
        'two',
        'three'
    ];

    public function updatedSelectedCheckboxes($value)
    {
       dump($value);
    }
}
```
7. edit /resources/views/livewire/test-checkboxes.blade.php
```
<div>
    @foreach($checkboxes as $checkbox)
        <input type="checkbox" value="{{ $checkbox }}" wire:model.defer="selectedCheckboxes"> {{ $checkbox }}
    @endforeach

    <button wire:click="$refresh">OK</button>
</div>

```
8. npm i && npm run build
9. php artisan serve
10. go to http://127.0.0.1:8000
11. uncheck checkbox two and click OK 

Livewire 3
1. laravel new lw3
2. cd lw3
3. composer require livewire/livewire:"^3.0"
4. php artisan livewire:make test-checkboxes
5. edit /resources/views/welcome.blade.php
```
<html>
    <head>
        @livewireScripts
    </head>
    <body>
        <livewire:test-checkboxes/>
    </body>
</html>

```
6. edit /app/Livewire/TestCheckboxes.php
```
<?php

namespace App\Livewire;

use Livewire\Component;

class TestCheckboxes extends Component
{
    public $checkboxes = [
        'one',
        'two',
        'three'
    ];

    public $selectedCheckboxes = [
        'one',
        'two',
        'three'
    ];

    public function updatedSelectedCheckboxes($value)
    {
       dump($value);
    }
}
```
7. edit /resources/views/livewire/test-checkboxes.blade.php
```
<div>
    @foreach($checkboxes as $checkbox)
        <input type="checkbox" value="{{ $checkbox }}" wire:model="selectedCheckboxes"> {{ $checkbox }}
    @endforeach

    <button wire:click="$refresh">OK</button>
</div>

```
8. npm i && npm run build
9. php artisan serve --port 8001
10. go to http://127.0.0.1:8001
11. uncheck checkbox two and click OK 

Notice the difference in the 2 dumps. In V2 the updated hook is called once with an array of the values, in V3 the updated hook is called for every value in the array, with a weird __rm__ value in the mix.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by luanfreitasdev** [comment]
I don't think it's a problem, because internally it's not setting the `__rm__`  property when we update the nested array, but it's confusing.

If I call `$search` and `updatedSearch()` will have `$value` set properly,
But if it's an array, `$checkbox` and call `updatedCheckbox()` inside a foreach, Livewire will do a recursion work and use `__rm__` to call unset().

```php
    public function updatedSelectedCheckboxes($value, $value2)
     {
         // after unchecked "three": $value = "__rm__", $value2 = 2 (index)
         // after checked "three": $value = "three", $value2 = 2 (index)
         dump($value, $value2);
     }
```
I don't know what the best approach is, maybe not using the `$value` parameter and using it inside the function: **$this->selectedCheckboxes** ?
Or update the documentation for these cases.

**Comment by joshhanley** [comment]
@AntonioDiPassio-AppSys thanks for reporting! Can you provide screenshots of both the dumps? Thanks!

**Reply by AntonioDiPassio-AppSys** [reply]
@joshhanley unfortunately I already removed both test installations, but I just tell you what they were:

in Livewire 2 the updated hook got called only once, with $value equal to array ['one','three']
in Livewire 3 the updated hook got called twice, with $value equal to string 'three' on the first iteration, and then $value was equal to string '__rm__' on the second iteration

**Comment by stebers** [comment]
I am experiencing also some weird behaviour of updating hook in combination with checkboxes in v3.

**Steps to Reproduce**

1. Add `public $checkboxes = []` property to a livewire component
2. Add 2 input fields to component's blade

- `<input type="checkbox" wire:model.live="checkboxes" value="box1" />`
-  `<input type="checkbox" wire:model.live="checkboxes" value="box2" />`

3. Add updating hook to component

- `public function updatingCheckboxes($value, $index) { if ($value !== '__rm__') return; dd($value, $index, $this->checkboxes); }`

4. Now test it in frontend by checking box1 and then box2. Then uncheck box1 while box2 is still checked and experience the following:
![image](https://github.com/livewire/livewire/assets/14885992/f7098409-9549-4319-bee4-4e958ed58b9b)

5. Note that `$this->checkboxes = ["box2", "box2"]` and that $index is not 0

My problem with this is that I would like to get the value in the updating hook by index like $this->checkboxes[$index], which is currently not working if there are more than one checkbox checked.

**Comment by seabasss** [comment]
I'm having the same problem with a multi select box. When items are removed, they get replaced by "__rm__", which makes it to the backend as well. I have to run this in the render method:

```
$this->form->tags = array_filter($this->form->tags, fn ($value) => $value !== '__rm__');
```

Had no issues in Live Wire 2. Any ideas?

**Comment by sha-hin** [comment]
Same problem, I've to move all updated logic when value is an array to the rendering method.
The hook gets called x times given x is the number of the array items.

**Comment by shokanshi** [comment]
Same issue with multi-select. Unselecting will cause the removed item to be replaced with "`__rm__`" and a form submit will cause this value to be sent to the server. 

Looking at the network from devtool, the following is observed under "updates" object:

`{selections.0: "__rm__", selections.1: "__rm__", selections.2: "three"}`

Forgot to mention, and because of this `__rm__ `, if all items are removed from the selection, the validation rule for `required` will still pass because the array is not empty.

**Comment by seabasss** [comment]
Any ideas on this topic? Thank you! :)

**Comment by jehadassaf** [comment]
This is a breaking change for our application, rendering it completely unusable and blocking any attempts at working around the issue and upgrading to v3. Was this change intentional or is this a bug? If this was intentional and is the new way forward, there is no way we can continue using Livewire because our application relies on many inputs with huge arrays. If this is a bug, can anyone with knowledge on the matter please have a look at it? I'd be glad to contribute and help but this is way beyond my knowledge of the Livewire internals and it's surely impacting a large number of users, many of whom may not even be aware of the issue as it's not documented in the upgrade guide.

**Comment by seabasss** [comment]
Any news on this topic? Thank you!

**Comment by iwan-izotov** [comment]
Hello, there is a hack. You can pass an object rather than an array in JavaScript and everything will work as in Liwevire2.
Example:
**Before**:

  ```
   treeSelectChangeHandler (values) {
        this.$wire.set('propertyIDs', values);
    },
```
 **After:** 
```
   treeSelectChangeHandler (values) {
        this.$wire.set('propertyIDs', Object.assign({}, values));
    },
```

**Reply by iamnoorsk** [reply]
it's work for me
thanks @iwan-izotov

**Comment by seabasseee** [comment]
Any news on this topic? Thank you!

**Comment by Xety** [comment]
This bug is still there in V3.4.9.

EDIT: Sorry to ping you @joshhanley but is that a confirmed bug or we do something bad ?
Cause i have to do the following to have my arrays working as intended :

```php
    /**
     * Function to update the model and return it after.
     *
     * @return Maintenance
     */
    public function update(): Maintenance
    {
        // Hack to counter a bug in Livewire
        // https://github.com/livewire/livewire/discussions/6665
        $this->operators = array_filter($this->operators, fn ($value) => $value !== '__rm__');
        $this->companies = array_filter($this->companies, fn ($value) => $value !== '__rm__');
        $this->incidents = array_filter($this->incidents, fn ($value) => $value !== '__rm__');

        //...
```

And i'm pretty sure that when I developped my website like 4/5 months ago, this bug wasn't there.

**Comment by shokanshi** [comment]
After much digging, I have finally found out the cause on my end for appending `__rm__` back into the result array after `updated` lifecycle hook. 

My component is using a custom trait that will trim strings upon any form submission and the trait is not taking into account of `__rm__` value and readding it back into the result after livewire passes through the `updated` hooks.

If you are facing this `__rm__` issue, check if you are using any custom trim string trait.

**Reply by seabasss** [reply]
Wow, that solved it for me too! Thanks for investigating and sharing! It's been bugging me for months.

**Reply by Xety** [reply]
I can confirm it was that for me too.

**Reply by joshhanley** [reply]
@shokanshi thanks for sharing! I had been looking into this and was wondering what was going on. Any chance you could share your trait before and after so others can see it? Thanks!

**Reply by shokanshi** [reply]
@joshhanley Sure. :)

Before:
```
trait TrimStrings
{
    public function updatedTrimStrings($name, $value)
    {
        if (is_string($value)) {
            $value = trim($value);
            data_set($this, $name, $value);
        }
    }
}
```

After:
```
trait TrimStrings
{
    public function updatedTrimStrings($name, $value)
    {
        if (is_string($value) && $value !== '__rm__') {
            $value = trim($value);
            data_set($this, $name, $value);
        }
    }
}
```

The trait is triggered from: 
```
foreach ($finishes as $finish) {
    $finish();
}
```
in the `updateProperties()` method of file: `HandleComponents.php`
```
protected function updateProperties($component, $updates, $data, $context)
{
    $finishes = [];

    foreach ($updates as $path => $value) {
        $value = $this->hydrateForUpdate($data, $path, $value, $context);

        // We only want to run "updated" hooks after all properties have
        // been updated so that each individual hook has the ability
        // to overwrite the updated states of other properties...
        $finishes[] = $this->updateProperty($component, $path, $value, $context);
    }

    foreach ($finishes as $finish) {
        $finish();
    }
}
```

**Reply by joshhanley** [reply]
@shokanshi thanks! 😁

**Reply by seabasss** [reply]
@joshhanley Here's my class and I added the `excludeTrimAndNullValues` array...

```
<?php

namespace App\Livewire\Concerns;

trait TrimAndNullEmptyStrings
{
    protected $trimAndNullEmptyStringsExcept = [
        //
    ];

    protected $excludeTrimAndNullValues = [
        '__rm__',
    ];

    public function updatedTrimAndNullEmptyStrings(string $name, $value): void
    {
        if (!is_string($value)) {
            return;
        }

        if (in_array($name, $this->trimAndNullEmptyStringsExcept) || in_array($value, $this->excludeTrimAndNullValues)) {
            return;
        }

        $trimmedValue = trim($value);
        $finalValue = $trimmedValue === '' ? null : $trimmedValue;

        data_set($this, $name, $finalValue);
    }
}
```

**Reply by Xety** [reply]
@joshhanley I have the same trait as @seabasss too.

Btw, why Livewire does not handle the "trim string" and "convert to null" by default like Laravel? It look like many people are using a custom trait to handle it as we can see in this discussion.
Can Livewire handle that by providing a default trait that we can include (or not) in our component/form ?

**Reply by seabasss** [reply]
@Xety His own words: https://github.com/livewire/livewire/issues/823#issuecomment-651324921

I think that Livewire is actually disabling the Laravel trim and null feature, so it’s not that it doesn’t support it.

**Reply by Xety** [reply]
@seabasss Yes but as you can see in this discussion and event in this discussion too https://github.com/livewire/livewire/issues/823#issuecomment-1487273304 people like to use those traits from Laravel. Btw the comment is from 3 years ago, so maybe his mind has changed now regarding this problem.
The real question is why disabling it in Livewire while Laravel use it by default ?
To avoid confusion I think Livewire should use it by default or at least provide a trait to use it if needed and document it.

**Reply by seabasss** [reply]
@Xety My guess is that Livewire is... live. So if you send data to the server on every keystroke you wouldn't be able to add spaces, since Livewire would trim them off all the time.

**Reply by joshhanley** [reply]
@seabasss Thanks for sharing!

@Xety yeah, @seabasss is correct. If we kept Laravel's `TrimStrings` and `ConvertEmptyStringsToNull` middleware, it would mess with the Livewire payload, which could trigger a `PayloadCorruptionException` due to the security check we have (as Livewire request properties are in a different structure than Laravel) and as mentioned, Livewire can have live data updates, so trimming a string could actually prevent a user from typing in a `space`.

We are slowly working on adding convert to null support to all of the property synths, so at least we don't end up with empty strings when something (like an int property) should be null instead.

But for trim strings you will still need to do that manually.

**Comment by adityadees** [comment]
What equivalent for this? 
In v2, I can see this working fine

        Livewire.hook('message.sent', (message, component) => {
            $("#responsive-datatable").DataTable().destroy()
        });

        Livewire.on("componentUpdate", (event) => {
          $('#responsive-datatable').DataTable();
        });
        
        
  but on v3 this not working

        Livewire.hook('commit', ({
            component,
        }) => {
            $("#responsive-datatable").DataTable().destroy();
        })
        
        Livewire.on('componentUpdate', () => {
            $("#responsive-datatable").DataTable();
        })

**Comment by shamil8124** [comment]
Pretty big issue for us here here too. Updated hook behaviour with checkboxes and multiselect became incredibly slow on a v3 upgrade.

**Comment by mauritskorse** [comment]
Maybe usefull for some. I have made a simple trait that determines what changes were made on the array and invokes new hooks: updatingAdd{Property}, updatingRemove{Property} and updatingReorder{Property}.
The trait also invokes these methods on updated hook, but it may not work as expected!

So if an item was removed, you can just use the updatingRemoveProperty hook which will only be called once, and not on every item in the array that had its key changed. If need to get those that had their key changed updatingReorderProperty will still be called for each item. updat**ed**OrderCollectionProperty will be called once and returns all the items that were reordered.

It works for my simple use case (short lists with only one item added or removed at a time). Haven't tested it in other scenarios.

```php
<?php
namespace App\Traits;

use Illuminate\Support\Arr;
use Illuminate\Support\Str;

/**
 * Livewire component trait for more detailed hooks on array data
 * Works as expected on updating hook, on updated hook it does not yet work as expected!
 */
trait HasArrayHook
{
    protected array $tmpArrayHookProperty;

    protected array $tmpArrayHookReorderCollection;

    /**
     * updatingHasArrayHook
     * @param array $params
     * @return void
     */
    public function updatingHasArrayHook(...$params)
    {
        // check if $params[1] is a nested array
        if(!Str::contains($params[0], '.')){
            return;
        }

        $this->arrayHookDetermineChanges('updating', ...$params);
    }
 
    /**
     * updatedHasArrayHook
     * @param array $params
     * @return void
     */
    public function updatedHasArrayHook(...$params)
    {
        // reset the temporary property
        $this->tmpArrayHookProperty['updating'] = null;
        
        if(!Str::contains($params[0], '.')){
            return;
        }
        $this->arrayHookDetermineChanges('updated', ...$params);
    }


    /**
     * renderingHasArrayHook
     * used to clear the temporary property
     * @return void
     */
    public function renderingHasArrayHook()
    {
        // reset the temporary property
        $this->tmpArrayHookProperty['updated'] = null;
    }
    

    /**
     * arrayHookDetermineChanges
     * Determines the actual changes in the array
     * Works as expecting on updating hook, on updated hook it does not yet work as expected
     * @param mixed $hook
     * @param array $params
     * @return void
     */
    public function arrayHookDetermineChanges($hook, ...$params){

        // get the property name of the nested array
        // since livewire returns a dot notated string of the key, we can use Laravel's Arr helper to get the value
        $dotNotated = Str::of($params[0]);
        $item = $params[1];

        $property = $dotNotated->before('.')->toString();
        $newItemKey = $dotNotated->afterLast('.')->toInteger();
        $itemPath = $dotNotated->after('.')->beforeLast('.')->toString();

        $oldArray = Arr::get($this->{$property}, $itemPath);
        $oldCount = is_array($oldArray) ? count($oldArray) : 0;

        // save old value (when multiple items)
        $this->tmpArrayHookProperty[$hook] = $this->tmpArrayHookProperty[$hook] ?? $oldArray;

        // check the cases: either reordering, removing or adding an item
        if(in_array($item, $oldArray) ){
            // get old kehy in nestedarra
            $oldKey = array_search($item, $oldArray);

            // implement new hook updating{$propery}Reordering: in case method exists, call it
            if(method_exists($this, $hook.'Reorder'.ucfirst($property))){
                $this->{$hook.'Reorder'.ucfirst($property)}($item, $oldKey, $newItemKey);
            }
            if($hook == 'updating'){
                $this->tmpArrayHookReorderCollection[$item] = ['old' => $oldKey, 'new' => $newItemKey];
            }
            if($hook == 'updated'){
                if($this->tmpArrayHookReorderCollection){
                    if(method_exists($this, $hook.'OrderCollection'.ucfirst($property))){
                        $this->{$hook.'OrderCollection'.ucfirst($property)}($this->tmpArrayHookReorderCollection);
                    }
                    $this->tmpArrayHookReorderCollection = [];
                }
            }
        }

        if($item == "__rm__"){
            // implement new hook updating{$propery}Removing: in case method exists, call it
            // when calling the updated hook, the old value is already removed so an empty array is passed
            if(method_exists($this, $hook.'Remove'.ucfirst($property))){
                $this->{$hook.'Remove'.ucfirst($property)}(array_diff($this->tmpArrayHookProperty[$hook], $oldArray));
            }
        }

        if($newItemKey >= $oldCount && $item != "__rm__"){
            // implement new hook updating{$propery}Adding: in case method exists, call it
            if(method_exists($this, $hook.'Add'.ucfirst($property))){
                $this->{$hook.'Add'.ucfirst($property)}($item);
            }
        }
    }
}
```

On your component you could do this:

```php
<?php


use App\Traits\HasArrayHook;
use Livewire\Component;

class MyComponent extends Component
{
    use HasArrayHook;

    public array $checkboxes;

    // your logic to populate the checkboxes

    public function updatingRemoveCheckboxes(...$params){
        dump('removing', $params);
    }

    public function updatingAddCheckboxes(...$params){
        dump('adding', $params);
    }
    public function updatingReorderCheckboxes(...$params){
        // each item that was reorderd invokes this hook
        dump('reordering', $params);
        $this->collectReorder[$params[0]] = ['old' => $params[1], 'new' => $params[2]];
    }

    public function updatedRemoveCheckboxes(...$params){
        // doesnt work yet as expected. is not invoked 
   }

    public function updatedAddCheckboxes(...$params){
        // doesnt work yet as expected. is not invoked
    }

    public function updatedReorderCheckboxes(...$params){
        dump('reordered', $params);
        // doesnt work yet as expected. Newly added or removed items will invoke Reorder and not Add or Remove respectively.
    }

    public function updatedOrderCollectionCheckboxes(...$params){
        dump('all reordered items', $params);
    }




```

**Comment by frntline-dev** [comment]
Any update on this issue? Seems to be a pretty big breaking change when migrating

**Comment by emaia** [comment]
Same problem here!

**Comment by afagard** [comment]
2 years later and this issue is still around...

**Reply by HolgerHatGarKeineNode** [reply]
We have cancelled all our Livewire 2 projects because of this change. It would have been far too much effort to migrate all the code.

The projects were then set up from scratch with Livewire 3 and the possibility to use new Laravel +11 functions and so on.

**Comment by RenoLooijmans** [comment]
Please also see https://github.com/livewire/livewire/pull/8235 and https://github.com/livewire/livewire/discussions/7101
Undesired side-effect of V3 is still around.

**Comment by RenoLooijmans** [comment]
Apparently this is going to be fixed in V4 with PR https://github.com/livewire/livewire/pull/9691.
